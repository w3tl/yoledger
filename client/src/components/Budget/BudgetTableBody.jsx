import React from 'react';
import PropTypes from 'prop-types';
import { accountPropType } from '../propTypes';
import BudgetTableColumn from './BudgetTableColumn';
import withData from './BudgetTableBodyHOC';
import { AccountInput } from '../Account';
import { Button } from '../Elements';

class BudgetTableBody extends React.Component {
  static propTypes = {
    Component: PropTypes.func,
    accounts: PropTypes.arrayOf(accountPropType),
    periods: PropTypes.arrayOf(PropTypes.string),
    onCreate: PropTypes.func,
    onPeriodChange: PropTypes.func,
  }

  static defaultProps = {
    Component: BudgetTableColumn,
    accounts: [],
    periods: [],
    onCreate: () => {},
    onPeriodChange: () => {},
  }

  constructor(props) {
    super(props);
    this.state = { newAccount: '' };
  }

  handlePeriodChange = sign => () => {
    const { onPeriodChange, periods } = this.props;
    if (sign > 0) {
      onPeriodChange(periods[1]);
    } else {
      const firstDate = new Date(periods[0]);
      firstDate.setDate(firstDate.getDate() - 1);
      onPeriodChange(firstDate.toISOString());
    }
  }

  handleNewAccountChange = (event) => {
    this.setState({ newAccount: event.target.value });
  }

  handleAdd = () => {
    const { newAccount: account } = this.state;
    const { periods, accounts, onCreate } = this.props;
    const existBudget = accounts.find(({ name }) => name === account);
    if (!existBudget) {
      onCreate({
        variables: {
          input: {
            date: periods[0] || new Date().toISOString(),
            account,
            amount: 0,
          },
        },
      });
    }
  }

  render() {
    const { newAccount = '' } = this.state;
    const { Component, accounts, periods } = this.props;
    return (
      <div style={{ display: 'flex' }}>
        <div>
          <div style={{ height: 30 }}>
            <Button id="prevButton" onClick={this.handlePeriodChange(-1)}>Prev</Button>
            <Button id="nextButton" onClick={this.handlePeriodChange(1)}>Next</Button>
          </div>
          {accounts.map(({ name }) => (
            <div style={{ height: 30 }} key={name}>{name}</div>
          ))}
          <div style={{ height: 30 }}>
            <AccountInput value={newAccount} onChange={this.handleNewAccountChange} />
            <Button disabled={newAccount.length === 0} id="add" onClick={this.handleAdd}>add</Button>
          </div>
        </div>
        {periods.map(date => (
          <Component key={date} accounts={accounts} date={date} />
        ))}
      </div>
    );
  }
}

export { BudgetTableBody as BudgetTableBodyComponent };
export default withData(BudgetTableBody);
