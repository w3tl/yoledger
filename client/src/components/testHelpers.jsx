import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

export const withProvider = Component => props => (
  <MockedProvider {...props}>
    <Component />
  </MockedProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return fixedDate;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
