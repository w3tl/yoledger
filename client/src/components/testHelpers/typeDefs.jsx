export default `
scalar Date
type Query {
  account(id: ID!): Account
  accounts(type: AccountType!): [Account]

  budgets(dateStart: Date!, count: Int = 5): Budget
  budget(date: Date!): [Allocation]

  transactions(dateStart: Date!, dateEnd: Date, page: Int = 0, itemsPerPage: Int = 20): [Transaction]
  transaction(id: ID!): Transaction
}
type Mutation {
  addAccount(input: AddAccountInput!): AddAccountPayload
  upsertBudget(input: UpsertBudgetInput!): UpsertBudgetPayload
  addTransaction(input: AddTransactionInput!): AddTransactionPayload
  deleteTransaction(id: ID!): DeleteTransactionPayload
  updateTransaction(id: ID!, input: UpdateTransactionInput!): UpdateTransactionPayload
}

type Account {
  id: ID!
  name: String!
  balance: Float
  type: AccountType!
}

type Budget {
  periods: [Date]
  accounts: [Account]
}

type Allocation {
  id: String!
  account: Account!
  amount: Float
  balance: Float
}

enum AccountType {
  ASSET
  EXPENSE
  INCOME
  VENDOR
}

input AddAccountInput {
  name: String!
  balance: Float
  type: AccountType!
}

type AddAccountPayload {
  account: Account
}

input UpsertBudgetInput {
  account: String!
  date: Date!
  amount: Float!
}

type UpsertBudgetPayload {
  success: Boolean
  allocation: Allocation
}

type Transaction {
  id: ID!
  amount: Float!
  source: Account!
  destination: Account!
  date: Date!
}
input AddTransactionInput {
  source: String!
  destination: String!
  amount: Float!
  date: Date
}
type AddTransactionPayload {
  transaction: Transaction
}
type DeleteTransactionPayload {
  success: Boolean!
  id: ID
}
input UpdateTransactionInput {
  amount: Float
  source: String
  destination: String
  date: Date
}
type UpdateTransactionPayload {
  transaction: Transaction
}
`;
