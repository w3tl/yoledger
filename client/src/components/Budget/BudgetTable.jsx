import React from 'react';
import PropTypes from 'prop-types';
import BudgetTableBody from './BudgetTableBody';

class BudgetTable extends React.Component {
  static propTypes = {
    body: PropTypes.func,
  }

  static defaultProps = {
    body: BudgetTableBody,
  }

  constructor(props) {
    super(props);
    const currentDate = new Date();
    currentDate.setDate(currentDate.getDate() - 1);
    this.state = {
      dateStart: currentDate.toISOString(),
      count: 6,
    };
  }

  handleHeaderClick = (dateStart) => {
    this.setState({
      dateStart,
    });
  }

  render() {
    const { body: Body } = this.props;
    const { dateStart, count } = this.state;

    return (
      <div>
        <Body dateStart={dateStart} count={count} onPeriodChange={this.handleHeaderClick} />
      </div>
    );
  }
}

export default BudgetTable;
