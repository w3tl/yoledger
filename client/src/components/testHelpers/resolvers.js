import dateResolver from '../../../../graphql/date';
import assets from '../Asset/mockData';
import { accounts, periods, budgets } from '../Budget/mockData';

export default {
  Date: dateResolver,
  Query: {
    account(root, { id }) {
      return assets.find(a => a.id === id);
    },
    accounts(root, { type }) {
      if (type === 'ASSET') return assets;
      return [];
    },
    budgets() {
      return { accounts, periods: periods.map(dateStr => new Date(dateStr)) };
    },
    budget() {
      return budgets[0];
    },
  },
  Mutation: {
    upsertBudget() {
      return {
        success: true,
        allocation: {
          account: { id: '99', name: 'new account' },
        },
      };
    },
  },
};
