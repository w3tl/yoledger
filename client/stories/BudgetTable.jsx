import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import BudgetTable from '../src/components/Budget/BudgetTable';
import { BudgetTableColumnComponent } from '../src/components/Budget/BudgetTableColumn';
import { BudgetTableBodyComponent } from '../src/components/Budget/BudgetTableBody';
import BudgetCell from '../src/components/Budget/BudgetCell';
import { accounts, periods, budgets } from '../src/components/Budget/mockData';

storiesOf('Budget/Cell', module)
  .add('empty', () => <BudgetCell onUpdate={action('onUpdate')} />)
  .add('with value', () => <BudgetCell amount={404} onUpdate={action('onUpdate')} />);

const ColumnWithProps = props => (
  <BudgetTableColumnComponent {...props} accounts={accounts} budget={budgets[0]} onUpdate={() => action('onUpdate')} />
);

const BodyWithProps = () => (
  <BudgetTableBodyComponent
    Component={ColumnWithProps}
    accounts={accounts}
    periods={periods}
    onCreate={action('onCreate')}
    onPeriodChange={action('onPeriodChange')}
  />
);

storiesOf('Budget/Table', module)
  .add('with budget', () => <BudgetTable body={BodyWithProps} />);
