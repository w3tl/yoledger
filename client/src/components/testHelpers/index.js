import React from 'react';
import { shallow, mount } from 'enzyme';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './typeDefs';
import resolvers from './resolvers';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

const executableSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

const cache = new InMemoryCache();
const link = new SchemaLink({ schema: executableSchema });
export const client = new ApolloClient({
  link,
  cache,
});

export const withProvider = Component => props => (
  <ApolloProvider client={client}>
    <Component client={client} />
  </ApolloProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  const date = new Date(fixedDate);
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return date;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
