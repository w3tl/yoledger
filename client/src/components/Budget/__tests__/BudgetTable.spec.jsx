import React from 'react';
import { shallow } from 'enzyme';
import BudgetTable from '../BudgetTable';
import { BudgetTableBodyComponent } from '../BudgetTableBody';
import { stubDate, testRenderWithoutError } from '../../testHelpers';

const fixedDate = new Date('2018-06-10');
describe('BudgetTable', () => {
  stubDate(fixedDate);

  testRenderWithoutError(
    <BudgetTable body={BudgetTableBodyComponent} />,
  );

  it('state should contain dateStart with current date and  default count', () => {
    const output = shallow(
      <BudgetTable body={BudgetTableBodyComponent} />,
    );
    const date = new Date(output.state('dateStart'));
    expect(date.getTime() < fixedDate.getTime()).toBeTruthy();
    expect(output.state('count')).toBeDefined();
  });
});
