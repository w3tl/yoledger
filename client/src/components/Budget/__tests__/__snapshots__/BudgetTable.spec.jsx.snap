// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BudgetTable renders without error 1`] = `
<div>
  <BudgetTableBody
    Component={[Function]}
    accounts={Array []}
    count={6}
    dateStart="2018-06-09T00:00:00.000Z"
    onCreate={[Function]}
    onPeriodChange={[Function]}
    periods={Array []}
  />
</div>
`;
